### PR TITLE
[smartswitch]: Fix DPU discovery from DPUS table

### PR DESCRIPTION
**Pull request by v-cshekar**
Support platforms that store DPU configuration in either the DPU or DPUS CONFIG_DB table.

Consolidate SmartSwitch gNOI upgrade parameters into one fixture and derive DPU indices consistently across the upgrade tests.

### Description of PR

  Fixes DPU discovery in the SmartSwitch gNOI upgrade test on platforms whose CONFIG_DB stores DPU config under the plural  DPUS   get_configured_dpu_names()  and  get_dpu_port()  only looked up  config_facts['DPU'] . On platforms using the  DPUS  table, that   lookup returned empty, so the upgrade fixture discovered zero DPUs and the test failed with a misleading gNOI error instead of  upgrading anything. This PR falls back to the  DPUS  table when  DPU  is absent.                                                     
                                                                                                                                       
   It also unifies the two overlapping SmartSwitch upgrade fixtures (gNOI-only and full NPU+DPU) into a single                         
 smartswitch_upgrade_params  fixture (a  SmartSwitchUpgradeParams  namedtuple) to reduce duplication.                                
   Where to start:  tests/common/platform/device_utils.py  (the fix) and  tests/upgrade_path/test_upgrade_smart_switch_gnoi.py     

Summary:
Fixes # (issue)

https://msazure.visualstudio.com/One/_workitems/edit/38896351/

### Type of change

- [x] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] New Test case
    - [ ] Skipped for non-supported platforms
- [ ] Test case improvement

### Back port request

- [ ] 202311
- [ ] 202405
- [ ] 202411
- [ ] 202505
- [ ] 202511
- [ ] 202512
- [x] 202605

Tracking issue/work item for backport/cherry-pick request (GitHub issue or Microsoft ADO):
Failure type: 

### Tested branch

- [ ] master
- [ ] 202311
- [ ] 202405
- [ ] 202411
- [ ] 202505
- [ ] 202511
- [ ] 202512
- [x] 202605
- [ ] N/A

### Test result

tests/upgrade_path/test_upgrade_smart_switch_gnoi.py::test_upgrade_smartswitch_all_dpus_then_npu[str3-cisco-8102-E04-U26]
tests/upgrade_path/test_upgrade_smart_switch_gnoi.py::test_upgrade_smartswitch_all_dpus_then_npu[str3-cisco-8102-E04-U26]
tests/upgrade_path/test_upgrade_smart_switch_gnoi.py::test_upgrade_smartswitch_all_dpus_then_npu[str3-cisco-8102-E04-U26]
tests/upgrade_path/test_upgrade_smart_switch_gnoi.py::test_upgrade_smartswitch_all_dpus_then_npu[str3-cisco-8102-E04-U26]
tests/upgrade_path/test_upgrade_smart_switch_gnoi.py::test_upgrade_smartswitch_all_dpus_then_npu[str3-cisco-8102-E04-U26]
tests/upgrade_path/test_upgrade_smart_switch_gnoi.py::test_upgrade_smartswitch_all_dpus_then_npu[str3-cisco-8102-E04-U26]
  /opt/venv/lib/python3.12/site-packages/ansible/plugins/loader.py:1485: UserWarning: AnsibleCollectionFinder has already been configured
    warnings.warn('AnsibleCollectionFinder has already been configured')

tests/upgrade_path/test_upgrade_smart_switch_gnoi.py: 145 warnings
  /usr/lib/python3.12/multiprocessing/popen_fork.py:66: DeprecationWarning: This process (pid=16159) is multi-threaded, use of fork() may lead to deadlocks in the child.
    self.pid = os.fork()

tests/upgrade_path/test_upgrade_smart_switch_gnoi.py::test_upgrade_smartswitch_all_dpus_then_npu[str3-cisco-8102-E04-U26]
  /var/src/sonic-mgmt-int/tests/conftest.py:1608: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
    record_testsuite_property("timestamp", datetime.utcnow())

tests/upgrade_path/test_upgrade_smart_switch_gnoi.py::test_upgrade_smartswitch_all_dpus_then_npu[str3-cisco-8102-E04-U26]
  /opt/venv/lib/python3.12/site-packages/passlib/utils/__init__.py:854: DeprecationWarning: 'crypt' is deprecated and slated for removal in Python 3.13
    from crypt import crypt as _crypt

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
- generated xml file: /var/src/sonic-mgmt-int/tests/logs/upgrade_path/test_upgrade_smart_switch_gnoi.py::test_upgrade_smartswitch_all_dpus_then_npu.xml -
---------------------------- live log sessionfinish ----------------------------
INFO     root:__init__.py:67 Can not get Allure report URL. Please check logs
DEBUG:tests.conftest:[log_custom_msg] item: <Function test_upgrade_smartswitch_all_dpus_then_npu[str3-cisco-8102-E04-U26]>
DEBUG:root:[Parallel Fixture] Terminate parallel manager after teardown
INFO:root:[Parallel Fixture] Terminating parallel fixture manager
DEBUG:root:[Parallel Fixture] Running tasks to be terminated: []
DEBUG:root:[Parallel Fixture] Current worker threads: []
DEBUG:root:[Parallel Fixture] Current worker threads: []
DEBUG:root:[Parallel Fixture] The fixture manager is terminated:
stopped tasks [],
canceled tasks [],
pending tasks [],
done tasks [].
INFO:root:Can not get Allure report URL. Please check logs
================= 1 passed, 159 warnings in 1501.31s (0:25:01) =================
EXIT_CODE=0

#### What is the motivation for this PR?
DPU upgrades were failing when DPU index was not specified.

#### How did you do it?
Fixed using the correct table.

#### How did you verify/test it?
For verification you can note: Validated on hardware             ┃
    str3-cisco-8102-E04-U26  —  test_upgrade_smartswitch_all_dpus_then_npu  upgraded all 8 DPUs  .31→.32  and the NPU to  .34 , passed  ┃
   (25 min).

### CODE REVIEW

File: tests/upgrade_path/test_upgrade_smart_switch_gnoi.py [hunk 1]
```diff
@@ -30,73 +32,83 @@ def _build_dpu_metadata(dpu_index: int):
...
         ss_target_index = 3
 
+    ss_max_workers = request.config.getoption("ss_max_workers")
+
     names = get_configured_dpu_names(duthost)
-    parsed_indices = None
-    if not ss_target_indices:
-        parsed_indices = list(range(len(names)))        # 4 -> [0, 1, 2, 3]]
+    ss_target_indices = request.config.getoption("ss_target_indices")
+    if names:
```
Comment (hdwhdw) [inline tests/upgrade_path/test_upgrade_smart_switch_gnoi.py]: Explicit `--ss_target_indices` is ignored whenever CONFIG_DB discovery returns any names. For example, `--ss_target_indices=2` on a four-DPU system produces `[0, 1, 2, 3]`, so this test can upgrade DPUs the operator did not select. Please parse the explicit option first and use discovery only when it is absent.

File: tests/upgrade_path/test_upgrade_smart_switch_gnoi.py [hunk 2]
```diff
@@ -30,73 +32,83 @@ def _build_dpu_metadata(dpu_index: int):
...
+        dpu_indices = list(range(len(names)))               # 4 -> [0, 1, 2, 3]
```
Comment (hdwhdw) [inline tests/upgrade_path/test_upgrade_smart_switch_gnoi.py]: `range(len(names))` assumes zero-based contiguous DPU IDs and discards the discovered identities. A table containing `dpu1` and `dpu3` would target DPUs 0 and 1; generated CONFIG_DB can also include `DPUS` entries for hardware that is admin-down. Please parse the actual indices and filter them using the enabled/admin-state source.

File: tests/common/platform/device_utils.py [hunk 3]
```diff
@@ -1311,7 +1311,7 @@ def get_dpu_port(duthost, dpu_index):
         logger.error("Failed to retrieve config_facts from DUT")
         return None
 
-    dpu_section = config_facts.get('DPU', {})
+    dpu_section = config_facts.get('DPU') or config_facts.get('DPUS') or {}
```
Comment (hdwhdw) [inline tests/common/platform/device_utils.py]: `DPUS` is a different schema, not an alternate spelling of `DPU`: SONiC defines `DPUS` as the name-to-midplane mapping, while `gnmi_port` belongs to `DPU`. On a standard `DPUS`-only configuration this lookup succeeds, but the subsequent `gnmi_port` lookup still returns `None`. Please keep inventory discovery separate from gNMI-port lookup.

**Review by hdwhdw — CHANGES_REQUESTED** [review]
TL;DR: The shared fixture can ignore explicit target selection and upgrade unintended DPUs. Please preserve explicit selector precedence, derive exact enabled IDs, and correct the `get_dpu_port` fallback; inline comments include failure scenarios.

Focused tests should cover explicit override, sparse or disabled IDs, and `DPUS`-only configuration.

File: tests/upgrade_path/test_upgrade_smart_switch_gnoi.py (hunk 1 above)
Comment (v-cshekar) [reply tests/upgrade_path/test_upgrade_smart_switch_gnoi.py]: Moved the logic to check `request.config.getoption("ss_target_indices")` first.  If the user has not provided the `--ss_target_indices` the all dpu's are selected.

File: tests/upgrade_path/test_upgrade_smart_switch_gnoi.py (hunk 2 above)
Comment (v-cshekar) [reply tests/upgrade_path/test_upgrade_smart_switch_gnoi.py]: Updated to use the actual indices and is filtered based on admin-state.

File: tests/common/platform/device_utils.py (hunk 3 above)
Comment (v-cshekar) [reply tests/common/platform/device_utils.py]: Added logic to check for `DPU` table.  If not present gnmi_port is checked.

**Review by hdwhdw — DISMISSED** [review]

**Comment by hdwhdw** [comment]
heading to @yxieca @StormLiangMS for merge

**Review by StormLiangMS — CHANGES_REQUESTED** [review]
@v-cshekar — ❌ Requesting changes because automatic DPU discovery currently fails open.

**Template**: ⚠️ `[x ]` is not valid checkbox syntax, and the tested branch/image details are not filled in.
**DCO**: ✅ All commits signed off.
**CI**: ✅ EasyCLA, static analysis, security, and KVM checks are green.

**Code Review:**
1. **[Critical] Missing admin state is treated as enabled.** `tests/common/platform/device_utils.py:1402` defaults a missing `CHASSIS_MODULE|DPU<index>.admin_status` to `up`. With a `DPUS`-only or partially populated configuration, automatic discovery can select unverified DPUs for a firmware upgrade. I reproduced `DPUS={dpu0,dpu3}` with no status table returning `[0, 3]`. The existing `upgrade_dpu_sonic_image` path instead fails on an invalid or missing admin status. Please fail closed here as well: skip/fail entries with missing state, use a reliable state fallback, or require explicit `--ss_target_indices` when state cannot be established.

Please address the above before merge.

**Comment by v-cshekar** [comment]
> @v-cshekar — ❌ Requesting changes because automatic DPU discovery currently fails open.
> 
> **Template**: ⚠️ `[x ]` is not valid checkbox syntax, and the tested branch/image details are not filled in. **DCO**: ✅ All commits signed off. **CI**: ✅ EasyCLA, static analysis, security, and KVM checks are green.
> 
> **Code Review:**
> 
> 1. **[Critical] Missing admin state is treated as enabled.** `tests/common/platform/device_utils.py:1402` defaults a missing `CHASSIS_MODULE|DPU<index>.admin_status` to `up`. With a `DPUS`-only or partially populated configuration, automatic discovery can select unverified DPUs for a firmware upgrade. I reproduced `DPUS={dpu0,dpu3}` with no status table returning `[0, 3]`. The existing `upgrade_dpu_sonic_image` path instead fails on an invalid or missing admin status. Please fail closed here as well: skip/fail entries with missing state, use a reliable state fallback, or require explicit `--ss_target_indices` when state cannot be established.
> 
> Please address the above before merge.

Made the change so that only DPUs with an explicit `CHASSIS_MODULE.admin_status`  of  up  are selected. Missing entries, missing   status fields, and unexpected status values are skipped

**Review by StormLiangMS — COMMENTED** [review]
@v-cshekar — ✅ The previously reported fail-open DPU-selection blocker is fixed on head `e91ee3706ea90158da45a26d1a220062817d9333`.

**Template**: ⚠️ Please change `[x ]` to `[x]`, select the tested branch(es), and provide the exact branch/image and evidence in **Test result**.
**DCO/CI**: ✅ DCO, EasyCLA, static/security checks, and impacted-area KVM jobs pass.

**Code Review:**
1. The revised discovery now preserves explicit `--ss_target_indices`, parses sparse real DPU IDs, keeps `DPU` gNMI-port lookup separate from `DPUS` inventory discovery, and selects automatically discovered DPUs only when `CHASSIS_MODULE.admin_status` is explicitly `up`.
2. Please add focused coverage for the edge cases that drove the earlier review: explicit override, sparse IDs, admin-down/missing status, and `DPUS`-only configuration.
3. The requested `202605` backport still lacks exact-scope closure evidence. ADO `38896351` reports no matching passing result for the original Mellanox-SN4280 test scope, while the PR documents a Cisco-8102 run. Please validate the original failing platform/scope or provide equivalent branch-specific evidence and justification.

The code blocker is resolved; the existing changes-requested review should remain until the template and validation gaps are addressed.

**Comment by v-cshekar** [comment]
> @v-cshekar — ✅ The previously reported fail-open DPU-selection blocker is fixed on head `e91ee3706ea90158da45a26d1a220062817d9333`.
> 
> **Template**: ⚠️ Please change `[x ]` to `[x]`, select the tested branch(es), and provide the exact branch/image and evidence in **Test result**. **DCO/CI**: ✅ DCO, EasyCLA, static/security checks, and impacted-area KVM jobs pass.
> 
> **Code Review:**
> 
> 1. The revised discovery now preserves explicit `--ss_target_indices`, parses sparse real DPU IDs, keeps `DPU` gNMI-port lookup separate from `DPUS` inventory discovery, and selects automatically discovered DPUs only when `CHASSIS_MODULE.admin_status` is explicitly `up`.
> 2. Please add focused coverage for the edge cases that drove the earlier review: explicit override, sparse IDs, admin-down/missing status, and `DPUS`-only configuration.
> 3. The requested `202605` backport still lacks exact-scope closure evidence. ADO `38896351` reports no matching passing result for the original Mellanox-SN4280 test scope, while the PR documents a Cisco-8102 run. Please validate the original failing platform/scope or provide equivalent branch-specific evidence and justification.
> 
> The code blocker is resolved; the existing changes-requested review should remain until the template and validation gaps are addressed.

@StormLiangMS  , have the review comments:

1.  The discovery now fails closed, a DPU is selected only when its admin_status is up. 
2. Added test cases covering all the cases: 

- Explicit override —  --ss_target_indices  wins; discovery not consulted

- Sparse IDs —  dpu1, dpu3 → [1, 3]  (parsed from real identities, not a positional count).   

-  Admin-down / missing status — admin-down skipped; missing  CHASSIS_MODULE  entry skipped; missing  admin_status  skipped;  include_admin_down=True  returns all. 

3. Cherry-picked the four fix commits onto  origin/internal-202605

```
● ✅ Test PASSED on the  internal-202605  branch.                                                                                                                                                             ┃┃                                              ┃
                                                                                                                                                                                                               ┃                                               ┃
   Result                                                                                                                                                                                                      ┃                                               ┃
                                                                                                                                                                                                               ┃                                               ┃
   ┌───────────────┬──────────────────────────────────────────────────────────────────────────────────┐                                                                                                        ┃                                               ┃
   │ Item          │ Value                                                                            │                                                                                                        ┃                                               ┃
   ├───────────────┼──────────────────────────────────────────────────────────────────────────────────┤                                                                                                        ┃                                               ┃
   │ Branch        │ v-cshekar/debug-dpu-202605 (cherry-picked onto origin/internal-202605)           │                                                                                                        ┃                                               ┃
   ├───────────────┼──────────────────────────────────────────────────────────────────────────────────┤                                                                                                        ┃                                               ┃
   │ Testbed       │ vms69-t1-smartswitch-8102-26 (DUT str3-cisco-8102-E04-U26)                       │                                                                                                        ┃                                               ┃
   ├───────────────┼──────────────────────────────────────────────────────────────────────────────────┤                                                                                                        ┃                                               ┃
   │ Test          │ test_upgrade_multiple_dpus_via_gnoi_parallel                                     │                                                                                                        ┃                                               ┃
   ├───────────────┼──────────────────────────────────────────────────────────────────────────────────┤                                                                                                        ┃                                               ┃
   │ Outcome       │ 1 passed in 308.90s (5:08), EXIT_CODE=0, core-dump/config check passed           │                                                                                                        ┃                                               ┃
   ├───────────────┼──────────────────────────────────────────────────────────────────────────────────┤                                                                                                        ┃                                               ┃
   │ DPU discovery │ dpu_indices=[0, 1, 2, 3, 4, 5, 6, 7] — all 8 auto-discovered by the fixed helper │                                                                                                        ┃                                               ┃
   ├───────────────┼──────────────────────────────────────────────────────────────────────────────────┤                                                                                                        ┃                                               ┃
   │ Images        │ DPU .31→.32 staged via gNOI SetPackage (DPU1–7 now Next=.32)                     │                                                                                                        ┃                                               ┃
   └───────────────┴──────────────────────────────────────────────────────────────────────────────────┘                                                                                                        ┃                                               ┃
                                                                                                                                                                                                               ┃                                               ┃
```

4. Updated this:  Please change `[x ]` to `[x]`.

**Comment by v-cshekar** [comment]
> This PR has backport request label(s) for branch(es): 202605, but is missing required test information. Please make sure you tick the tested branch(es) in the **Tested branch** section and provide test evidence (e.g., 202605: <test result>) in the **Test result** section as well in your PR description.
> 
> ---Powered by SONiC BuildBot

Done selected the tested branch and update the test results in PR description as well.

**Review by StormLiangMS — COMMENTED** [review]
## PR Review — `[smartswitch]: Fix DPU discovery from DPUS table`

### ✅ What looks good

1. **Core fix is correct** — `get_configured_dpu_names()` now falls back: `config_facts.get('DPU') or config_facts.get('DPUS') or {}`. Directly fixes the zero-DPU discovery bug.
2. **`get_dpu_port()` intentionally does NOT fallback to DPUS** — makes sense since `gnmi_port` only exists in the `DPU` table. Well-documented.
3. **Hostname-prefixed key matching** — `re.search(r'(\d+)$', name)` handles both `dpu0` and `<host>-dpu0` style keys.
4. **`get_configured_dpu_indices()` is well-designed** — fail-closed admin-state filtering is the safe default for firmware upgrade scenarios.
5. **Fixture consolidation** — merging two overlapping fixtures into `SmartSwitchUpgradeParams` namedtuple eliminates duplication.
6. **Unit tests** — 13 test cases with good coverage of edge cases.

### ⚠️ Issues / Suggestions

**1. Double `config_facts` call in `get_configured_dpu_indices()` (Medium)**

`get_configured_dpu_names(duthost)` calls `config_facts` internally, then `get_configured_dpu_indices()` calls it again. This hits the DUT twice for the same data. Consider refactoring `get_configured_dpu_names()` to accept pre-fetched config_facts, or extract a shared internal helper.

**2. Trailing-digit regex may over-match (Low)**

`re.search(r'(\d+)$', str(name))` — a key like `something123` (no "dpu") would match `123`. Consider anchoring: `r'dpu(\d+)$'` (case-insensitive) to be more precise.

**3. Skip message in `test_upgrade_multiple_dpus_via_gnoi_parallel` is misleading (Low)**

Now that `dpu_indices` comes from auto-discovery, the skip message `"--ss-target-indices not set; ..."` is incomplete. Consider: `"No DPU indices found (no admin-up DPUs discovered and --ss_target_indices not set)"`.

**4. `ss_target_index` default of `3` — rationale unclear (Low)**

On a platform with only `dpu0, dpu1`, defaulting to `3` would silently target a non-existent DPU. A comment explaining the choice or defaulting to the first discovered index would help.

**Review by StormLiangMS — APPROVED** [review]

**Comment by StormLiangMS** [comment]
> ## PR Review — `[smartswitch]: Fix DPU discovery from DPUS table`
> ### ✅ What looks good
> 1. **Core fix is correct** — `get_configured_dpu_names()` now falls back: `config_facts.get('DPU') or config_facts.get('DPUS') or {}`. Directly fixes the zero-DPU discovery bug.
> 2. **`get_dpu_port()` intentionally does NOT fallback to DPUS** — makes sense since `gnmi_port` only exists in the `DPU` table. Well-documented.
> 3. **Hostname-prefixed key matching** — `re.search(r'(\d+)$', name)` handles both `dpu0` and `<host>-dpu0` style keys.
> 4. **`get_configured_dpu_indices()` is well-designed** — fail-closed admin-state filtering is the safe default for firmware upgrade scenarios.
> 5. **Fixture consolidation** — merging two overlapping fixtures into `SmartSwitchUpgradeParams` namedtuple eliminates duplication.
> 6. **Unit tests** — 13 test cases with good coverage of edge cases.
> 
> ### ⚠️ Issues / Suggestions
> **1. Double `config_facts` call in `get_configured_dpu_indices()` (Medium)**
> 
> `get_configured_dpu_names(duthost)` calls `config_facts` internally, then `get_configured_dpu_indices()` calls it again. This hits the DUT twice for the same data. Consider refactoring `get_configured_dpu_names()` to accept pre-fetched config_facts, or extract a shared internal helper.
> 
> **2. Trailing-digit regex may over-match (Low)**
> 
> `re.search(r'(\d+)$', str(name))` — a key like `something123` (no "dpu") would match `123`. Consider anchoring: `r'dpu(\d+)$'` (case-insensitive) to be more precise.
> 
> **3. Skip message in `test_upgrade_multiple_dpus_via_gnoi_parallel` is misleading (Low)**
> 
> Now that `dpu_indices` comes from auto-discovery, the skip message `"--ss-target-indices not set; ..."` is incomplete. Consider: `"No DPU indices found (no admin-up DPUs discovered and --ss_target_indices not set)"`.
> 
> **4. `ss_target_index` default of `3` — rationale unclear (Low)**
> 
> On a platform with only `dpu0, dpu1`, defaulting to `3` would silently target a non-existent DPU. A comment explaining the choice or defaulting to the first discovered index would help.

hi @v-cshekar pls take the suggestion as follow up.
